### PR TITLE
fix(NODE-4735): fix change stream consecutive resumabilty

### PR DESCRIPTION
**Pull request by baileympearson**
#### What is changing?

Change streams should loop until a document is produced, so that in the case of a successful resume, a subsequent resumable error is treated as its own resumable error.

##### Is there new documentation needed for these changes?

No.

#### What is the motivation for this change?

Correct change stream resume behavior.

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

File: src/change_stream.ts [hunk 1]
```diff
@@ -648,21 +648,21 @@ export class ChangeStream<
   hasNext(callback?: Callback): Promise<boolean> | void {
     this._setIsIterator();
     return maybeCallback(async () => {
-      try {
-        const hasNext = await this.cursor.hasNext();
-        return hasNext;
-      } catch (error) {
+      for (;;) {
```
Comment (baileympearson) [inline src/change_stream.ts:651]:
```suggestion
      while(true) {
```

File: src/change_stream.ts [hunk 2]
```diff
@@ -675,23 +675,22 @@ export class ChangeStream<
   next(callback?: Callback<TChange>): Promise<TChange> | void {
     this._setIsIterator();
     return maybeCallback(async () => {
-      try {
-        const change = await this.cursor.next();
-        const processedChange = this._processChange(change ?? null);
-        return processedChange;
-      } catch (error) {
+      for (;;) {
```
Comment (baileympearson) [inline src/change_stream.ts:678]:
```suggestion
      while(true) {
```

File: src/change_stream.ts [hunk 3]
```diff
@@ -706,21 +705,21 @@ export class ChangeStream<
   tryNext(callback?: Callback<Document | null>): Promise<Document | null> | void {
     this._setIsIterator();
     return maybeCallback(async () => {
-      try {
-        const change = await this.cursor.tryNext();
-        return change ?? null;
-      } catch (error) {
+      for (;;) {
```
Comment (baileympearson) [inline src/change_stream.ts:708]:
```suggestion
      while(true) {
```

File: src/change_stream.ts (hunk 1 above)
Comment (dariakp) [inline src/change_stream.ts:651]: This is less obvious than a while true loop, so unless there's a particularly good reason to use this, I'd prefer the while loop (yes even with the lint warning, maybe even with a comment describing why this is definitely intentional)

File: test/integration/change-streams/change_stream.test.ts [hunk 4]
```diff
@@ -1767,7 +1718,44 @@ describe('ChangeStream resumability', function () {
             expect(aggregateEvents).to.have.lengthOf(2);
           }
         );
+
+        it(
+          `supports consecutive resumes on error code ${code} ${error}`,
```
Comment (dariakp) [inline test/integration/change-streams/change_stream.test.ts:1723]: Just a note that these test additions will likely result in merge conflicts for https://github.com/mongodb/node-mongodb-native/pull/3454 and we should be extra careful in that PR, once it's rebased, that all these tests are pulled in

File: src/change_stream.ts [hunk 5]
```diff
@@ -648,21 +648,25 @@ export class ChangeStream<
   hasNext(callback?: Callback): Promise<boolean> | void {
     this._setIsIterator();
     return maybeCallback(async () => {
-      try {
-        const hasNext = await this.cursor.hasNext();
-        return hasNext;
-      } catch (error) {
+      // Change streams must resume indefinitely as long while each resume event succeeds.
```
Comment (dariakp) [inline src/change_stream.ts:651]: Not sure the best way to fix "as long while", but it doesn't work semantically at the moment

**Review by dariakp — CHANGES_REQUESTED** [review]

**Review by dariakp — APPROVED** [review]

**Review by nbbeeken — APPROVED** [review]
LGTM

**Pull request merged**